bandit/bandit.go

```diff
@@ -72,77 +95,66 @@ func (o *BanditDialer) DialContext(ctx context.Context, network, addr string) (n
-		// the normalized raw throughput to determine the reward. This is because the
-		// reward system takes into account how many tries there have been for a given
-		// "arm", so giving a reward here would be double-counting.
-
-		// Tell the dialer to update the bandit with it's throughput after 5 seconds.
-		dt := newDataTrackingConn(conn)
-		time.AfterFunc(secondsForSample*time.Second, func() {
-			speed := normalizeReceiveSpeed(dt.dataRecv)
-			//log.Debugf("Dialer %v received %v bytes in %v seconds, normalized speed: %v", dialer.Name(), dt.dataRecv, secondsForSample, speed)
-			o.bandit.Update(chosenArm, speed)
-		})
-		o.onSuccess(dialer)
-		return dt, err
+		return nil, err
 	}
-	o.onFailure()
-	return nil, &BanditError{
-		attempts: len(o.dialers),
+	log.Debugf("Dialer %v dialed in %v seconds", dialer.Name(), time.Since(start).Seconds())
+	// We don't give any special reward for a successful dial here and just rely on
+	// the normalized raw throughput to determine the reward. This is because the
+	// reward system takes into account how many tries there have been for a given
+	// "arm", so giving a reward here would be double-counting.
+
+	// Tell the dialer to update the bandit with it's throughput after 5 seconds.
+	dt := newDataTrackingConn(conn)
+	time.AfterFunc(secondsForSample*time.Second, func() {
+		speed := normalizeReceiveSpeed(dt.dataRecv)
+		//log.Debugf("Dialer %v received %v bytes in %v seconds, normalized speed: %v", dialer.Name(), dt.dataRecv, secondsForSample, speed)
+		o.bandit.Update(chosenArm, speed)
+	})
+	o.onSuccess(dialer)
+	return dt, err
+}
+
+func (o *BanditDialer) chooseDialerForDomain(dialers []Dialer, network, addr string) (Dialer, int) {
+	chosenArm := o.bandit.SelectArm(rand.Float64())
+	dialer := dialers[chosenArm]
+	if dialer.SupportsAddr(network, addr) {
+		return dialer, chosenArm
 	}
+	chosenArm = differentArm(chosenArm, len(dialers), o.bandit)
+	return dialers[chosenArm], chosenArm
 }
 
 // Choose a different arm than the one we already have, if possible.
 func differentArm(existingArm, numDialers int, eg *bandit.EpsilonGreedy) int {
-	// We let the bandit choose a new arm (or at least try to) versus just rotating to the next
-	// dialer because we want to use the bandit's algorithm for optimizing exploration
-	// versus exploitation, i.e. it will choose another dialer with a probability
-	// proportional to how well it has performed in the past as well as to whether
-	// or not we need more data from it. Basically, it will choose a more useful
-	// dialer than our random selection.
-	if numDialers == 1 {
-		log.Debugf("Only one dialer, so returning the existing arm: %v", existingArm)
-		return existingArm
-	}
+	// We try to choose a different arm randomly.
 	for i := 0; i < 20; i++ {
-		newArm := eg.SelectArm(rand.Float64())
+		// This selects a new arm randomly with protobability 1.0
+		// of choosing exploration vs exploitation.
+		newArm := eg.SelectArm(1.0)
```

Comment:
Won't this always return the same arm...?
> SelectArm chooses an arm that exploits if the value is more than the epsilon
threshold, and explore if the value is less than epsilon

Am I reading that wrong?

Reply:
Seems like yes?  (`max` is deterministic)
```
func (b *EpsilonGreedy) SelectArm(probability float64) int {
	b.RLock()
	defer b.RUnlock()

	// Exploit
	if probability > b.Epsilon {
		return max(b.Rewards...)
	}

	// Explore
	return rand.Intn(len(b.Rewards))
}
```

Reply:
Ah right I wanted to go in the other direction (low probability of exploitation). Great catch!